lib/project/setup-project.js

```diff
@@ -2,32 +2,65 @@
 const semverCoerce = require('semver/functions/coerce');
 const YarnLockfile = require('@yarnpkg/lockfile');
 const npa = require('npm-package-arg');
+const ini = require('ini');
 const fs = require('fs');
 const path = require('path');
 const debug = require('debug')('supported:project');
+const { setupProjectPath } = require('../read-from-url');
 
 const npmConfig = require('../npm/config');
 
-module.exports = async function setupProject(projectRoot) {
-  const config = await npmConfig(projectRoot); // kinda slow, TODO: re-implement as standalone lib
-  // const { policies } = options;
-  const pkgPath = `${projectRoot}/package.json`;
-  if (!fs.existsSync(pkgPath)) {
-    throw new Error(`${pkgPath} does not exist, are you sure this is a valid package?`);
-  }
-  if (!fs.statSync(pkgPath).isFile()) {
-    throw new Error(`${pkgPath} is not a file, are you sure this is a valid package?`);
+module.exports = async function setupProject(projectRoot, flags = {}) {
+  let projectInfo = {};
+  if (!fs.existsSync(projectRoot)) {
+    projectInfo = await setupProjectPath(projectRoot, {
+      token: flags.token,
+      hostUrl: flags.hostUrl,
+    });
+    // if project root is a URL then use homedir as root
+    projectRoot = require('os').homedir();
   }
-  const file = fs.readFileSync(pkgPath, 'utf-8');
-  let pkg;
-  try {
-    pkg = JSON.parse(file);
-  } catch (e) {
-    throw new Error(`${pkgPath} is not a valid JSON file, are you sure this is a valid package?`);
+
+  let config = await npmConfig(projectRoot); // kinda slow, TODO: re-implement as standalone lib
+  let pkg = {};
+  let lockfileContent = '';
+  if (projectInfo.packageJSON) {
+    pkg = projectInfo.packageJSON;
+    lockfileContent = projectInfo['yarn.lock'];
+    if (projectInfo['.npmrc']) {
+      try {
+        let localConfig = ini.parse(projectInfo['.npmrc']);
+        config = {
+          config,
+          ...localConfig,
+        };
+      } catch (e) {
+        throw new Error(
```

Comment:
I'm always nervous about catching all errors, is there a way to detect the expected exception if there was a parse failure and provide a nicer error? (assuming the existing error isn't itself already nice)

Reply:
We can append the actual error this error so that we are not hiding actual error.

Reply:
In the new change, I have appended the actual error.
I personally didn't see any error. I have just put the try-catch just to put a nice error message before we throw the actual error stack.